shared/middleware/helpers/bucket_helper.rb

```diff
@@ -206,12 +206,21 @@ def check_current_version(encrypted_channel_id, filename, current_version, shoul
     owner_id, channel_id = storage_decrypt_channel_id(encrypted_channel_id)
     key = s3_path owner_id, channel_id, filename
 
-    # This is an Array of ObjectVersions, defined in:
-    # https://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Types/ObjectVersion.html
-    track_list_operation 'BucketHelper.check_current_version'
-    versions = s3.list_object_versions(bucket: @bucket, prefix: key).versions
+    expected_latest_version = begin
+      s3.head_object(bucket: @bucket, key: key, version_id: current_version).version_id
+    rescue Aws::S3::Errors::NoSuchKey, Aws::S3::Errors::NoSuchVersion, Aws::S3::Errors::InvalidArgument
+      nil # No main.json yet, invalid version id or version not found
+    end
 
-    target_version_metadata = versions.find {|v| v.version_id == current_version}
+    # This second request can be skipped if we encountered an S3 eventual consistency issue.
+    actual_latest_version = nil
+    if expected_latest_version
+      begin
+        actual_latest_version = s3.head_object(bucket: @bucket, key: key).version_id
+      rescue Aws::S3::Errors::NoSuchKey
+        # No main.json yet
+      end
+    end
```

Comment:
Realized after asking for review that it _might_ be more optimal to early-out the other way around.

As is, I ask for the version the _client_ thinks is current.  If we run into an S3 eventual consistency issue (rare?) this will be nil, and we can skip the second request for the actual current version.

We could ask for the actual current version first, and if it matches what the client this is current (common), then we implicitly know the expected latest version exists, and can skip looking up the expected latest version.